pop_tools/zonal_average.py

```diff
@@ -0,0 +1,247 @@
+    )
+
+    # Merge datasets with data and grid corner information
+    out_ds = xr.merge([data_ds, grid_corners])
+
+    return out_ds
+
+
+def _generate_weights(src_grid, dst_grid, method, weight_file=None):
+    """
+    Generate regridding weights by calling xESMF
+    """
+
+    # Allow user to input weights file, if there is not one, use default check
+    if weight_file is None:
+        weight_file = _get_default_filename(src_grid, dst_grid, method)
+
+    # Check to see if the weights file already exists - if not, generate weights
+    if not os.path.exists(weight_file):
+        regridder = xe.Regridder(src_grid, dst_grid, method)
+        print('Saving weights file...')
+        regridder.to_netcdf(weight_file)
+
+    else:
+        regridder = xe.Regridder(src_grid, dst_grid, method, weights=weight_file)
+
+    return regridder
+
+
+class Regridder:
+    def __init__(
+        self,
+        grid_name=None,
+        grid=None,
+        dx=None,
+        dy=None,
+        mask=None,
+        regrid_method='conservative',
+        method_gen_grid='regular_lat_lon',
+    ):
+        """
+        A regridding class which uses xESMF and Xarray tools to both regrid and
+        calculate a zonal averge.
+
+        Parameters
+        ----------
+        grid_name: string
+          POP grid name (ex. 'POP_gx1v6')
+
+        grid: `xarray.Dataset`
+          User defined grid containing metadata typically found in POP grid
+
+        dx: float
+          Horizontal grid spacing in x-direction for output grid in degrees
+
+        dy: float
+          Horizontal grid spacing in y-direction for output grid in degrees
+
+        mask: `xarray.Dataarray`
+          User defined region mask
+
+        regrid_method: string
+          Regridding method to be used within xESMF (default is conservative)
+
+        method_gen_grid: string
+          Method used to generate the output grid - default is a regular lat/lon grid
+        """
+        if grid_name is not None:
+            self.grid_name = grid_name
+
+            # Use pop-tools to retrieve the grid
+            self.grid = get_grid(grid_name)
+
+        elif grid is not None:
+            self.grid = grid
+
+        else:
+            raise ValueError('Failed to input grid name or grid dataset')
+
+        # Set the dx/dy parameters for generating the grid
+        self.dx = dx
+        self.dy = dy
+
+        # Set the regridding method
+        self.regrid_method = regrid_method
+
+        # Set the grid generation method
+        self.method_gen_grid = method_gen_grid
+
+        # If the user does not input a mask, use default mask
+        if not mask:
+            self.mask = self.grid['REGION_MASK']
+
+            if 'region_name' not in self.grid.variables:
+                self.mask_labels = np.unique(self.mask.values)
+
+            else:
+                self.mask_labels = self.grid['region_name']
+
+        else:
```

Comment:
Do we never need no mask? I guess a global average across all basins is somewhat useless but would be required for the atmosphere.